src/axom/quest/SamplingShaper.hpp

```diff
@@ -108,10 +116,38 @@ class InOutSampler
     m_octree->generateIndex();
   }
 
-  void sampleInOutField(mfem::DataCollection* dc,
-                        shaping::QFunctionCollection& inoutQFuncs,
-                        int sampleRes)
+  /**
+   * \brief Samples the inout field over the indexed geometry, possibly using a
+   * callback function to project the input points (from the computational mesh)
+   * to query points on the spatial index
+   * 
+   * \tparam FromDim The dimension of points from the input mesh
+   * \tparam ToDim The dimension of points on the indexed shape
+   * \param [in] dc The data collection containing the mesh and associated query points
+   * \param [inout] inoutQFuncs A collection of quadrature functions for the shape and material
+   * inout samples
+   * \param [in] sampleRes The quadrature order at which to sample the inout field
+   * \param [in] projector A callback function to apply to points from the input mesh
+   * before querying them on the spatial index
+   * 
+   * \note A projector callback must be supplied when \a FromDim is not equal 
+   * to \a ToDim, the projector
+   * \note \a ToDim must be equal to \a DIM, the dimension of the spatial index
+   */
+  template <int FromDim, int ToDim = DIM>
```

Comment:
Kenny, this may be a style question more than anything, but--- why do you have `ToDim`, `DIM`, and `NDIMS`?  

First, why have `ToDim` at all?  You're already templating on destination space dimension, and disabling `sampleInOutField` if `ToDim` isn't the right value.  Isn't that writing a pitfall for a user?

Second, the class `InOutSampler` is templated on `NDIMS`.  Why do you have `static constexpr int DIM = NDIMS;` at (new) line 65?  Is that because you needed to use that value in the code, and pulling it into a variable is the most convenient way to do it?  (If so, I guess that makes sense.)

Reply:
Great question -- lots to unpack here:
* the class needs a template parameter on the dimension since `primal` and `spin` classes are templated by dimension. I used `NDIMS` for that parameter since that appears to be the de-facto standard for this in axom, see, e.g. `primal::Point` and `spin::BVH`
* As you mentioned in the second point above, we often need to figure out the dimension of a class that we're using. Unfortunately, you can't directly access the template parameter from other classes and compilers complain when you redefine the same name, so we expose that as a static constexpr int -- `DIM`.  Looks like there are some inconsistencies on this convention in axom, e.g. `Point` has `DIMENSION`, but `Triangle` has `DIM`
* For this function in particular, since we're taking a lambda for input points in specific, but possibly different dimension that projects to query points in another specific dimension, we need an additional template parameter, which I've (somewhat arbitrarily) called `FromDim` 
* I was hoping to not have to define a `ToDim` parameter here, since, as you've rightly pointed out, the code is only valid when `ToDim == DIM == NDIMS`. Unfortunately, the compiler complained when I removed it. (Please see my comment on the  `ToDim != DIM` overload a few lines down. The good news is that we don't have to supply this parameter at the calling site due to the default value.
    * If anyone knows of a way to remove the `ToDim` parameter and the extra function while keeping the code compiling, I'd happily remove it.

Some links:
* https://github.com/LLNL/axom/blob/develop/src/axom/primal/geometry/Point.hpp
* https://github.com/LLNL/axom/blob/develop/src/axom/spin/BVH.hpp
* https://github.com/LLNL/axom/blob/b233024f2baf616ee99a6a5299781773fa45edbb/src/axom/primal/geometry/Point.hpp#L66
* https://github.com/LLNL/axom/blob/b233024f2baf616ee99a6a5299781773fa45edbb/src/axom/primal/geometry/Triangle.hpp#L52

Reply:
There isn't a way to do it with std::enable_if, where the case ToDim != DIM emits a warning message? Or is that too annoying?

Reply:
@kennyweiss , thanks for the detailed comment.  That does all make sense, and confirms what I thought.